File: ADSAA/To_Fill_or_Not_to_Fill .cpp
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

// define infinty
const int inf = 99999999;
struct station {
    double price, dis;
};
bool cmp1(station a, station b) {
    return a.dis < b.dis;
}
int main() {
    //get input
    double cmax, d, davg;
    int n;
    scanf("%lf%lf%lf%d", &cmax, &d, &davg, &n);
    
    //using strucure vector
    vector<station> sta(n + 1); //STL point
    sta[0] = {0.0, d};
    for(int i = 1; i <= n; i++)
        scanf("%lf%lf", &sta[i].price, &sta[i].dis);
    sort(sta.begin(), sta.end(), cmp1); //STL point
    
    //Initial, get start, leftdis=how far can go with gas left in tank
    double nowdis = 0.0, maxdis = 0.0, nowprice = 0.0, totalPrice = 0.0, leftdis = 0.0;
    if(sta[0].dis != 0) {
        printf("The maximum travel distance = 0.00"); // no station at begin, nowhere can go
        return 0;
    } else {
        nowprice = sta[0].price;
    }

    //enter loop
    while(nowdis < d) {
        maxdis = nowdis + cmax * davg;
        double minPriceDis = 0, minPrice = inf;
        int flag = 0;
        //search farther station in now max distance 
        for(int i = 1; i <= n && sta[i].dis <= maxdis; i++) {
            if(sta[i].dis <= nowdis) continue; // not farther, neglect, jump out
            //find lower price, add gas at this station, just enough reach goal station
            if(sta[i].price < nowprice) {
                totalPrice += (sta[i].dis - nowdis - leftdis) * nowprice / davg; 
                leftdis = 0.0;
                nowprice = sta[i].price;
                nowdis = sta[i].dis;
                flag = 1; //find
                break;
            }
            //update minPrice
            if(sta[i].price < minPrice) {
                minPrice = sta[i].price;
                minPriceDis = sta[i].dis;
            }
        }
        // not found, but exist farther station
        if(flag == 0 && minPrice != inf) {
            totalPrice += (nowprice * (cmax - leftdis / davg));
            leftdis = cmax * davg - (minPriceDis - nowdis);
            nowprice = minPrice;
            nowdis = minPriceDis;     
        }
        // not found, but no farther station
        if(flag == 0 && minPrice == inf) {
            nowdis += cmax * davg; //drive with full gas from this station
            printf("The maximum travel distance = %.2f", nowdis);
            return 0;
        }
    }
    printf("%.2f", totalPrice);
    return 0;
}

File: ADSAA/To_Fill_or_Not_to_Fill .py
# https://qsctech-sange.github.io/1033-To-Fill-or-Not-to-Fill.html#sample-output-2
def next_station(station, now_tank_can_run, price):
    global num_sta, stations, tank, max_dis, per_unit, distance
    if station == num_sta:
        return True, price
    # next_sta 先是下一个车站
    next_sta = station + 1
    
    # 如果下一个车站到这个车站的距离超过了最大能开的距离，报错
    if stations[next_sta][1] - stations[station][1] > max_dis:
        return False, stations[station][1] + max_dis
    
    # 否则，next_sta 找到下一个要前去的车站，应当是油价比现在低的第一个车站。'先遍历里程范围内的第一个油价比此车站低的车站'
    # ‘离开这个循环可能是因为next_sta油价<station油价，或者 next_sta 超出当前station的最大距离’
    while stations[next_sta][0] > stations[station][0] and stations[next_sta + 1][1] - stations[station][1] <= max_dis:
        next_sta += 1
    
    # 如果next_sta还是比现在的价格高，说明里程范围内没有比现在更低的油价，'那么就把现在的邮箱拉满，同时找到下一个里程范围内价格相对更低的车站'
    if stations[next_sta][0] > stations[station][0]:
        next_sta = station + 1
        temp = next_sta
        while stations[temp + 1][1] - stations[station][1] <= max_dis:
            temp += 1
            if stations[temp][0] < stations[next_sta][0]:
                next_sta = temp
        return next_station(next_sta, max_dis - stations[next_sta][1] + stations[station][1],
                            price + (max_dis - now_tank_can_run) / per_unit * stations[station][0])
    
    # 否则，next_sta油价更低，
    # 正好移动到next_station为止
    else:
        return next_station(next_sta, 0,
                            price + (stations[next_sta][1] - stations[station][1] - now_tank_can_run) / per_unit *
                            stations[station][0])

# get input
tank, distance, per_unit, num_sta = list(map(float, input().split()))
max_dis = tank * per_unit
stations = []

# initial
for _ in range(int(num_sta)):
    info = input().split()
    stations.append([float(info[0]), float(info[1])])

# add destation as a station with 0 price
stations.append([0, distance])

# sort by distance
stations.sort(key=lambda x: x[1])


if stations[0][1] != 0: # no station at begin
    print("The maximum travel distance = 0.00")
else:
    status, price = next_station(0, 0, 0)
    if status:
        print("%0.2f" % price)
    else:
        print("The maximum travel distance = %0.2f" % price)